### ISSUE TITLE
Presence of links cause unexpected line breaks in paragraphs

### ISSUE THREAD

**Issue by oblitum** [lunaryorn/mdcat#181]
```
❯ cat foo.md
foo
[baz](https://baz.baz)
bar

❯ mdcat foo.md
foo
baz[1]
bar

[1]: https://baz.baz
```

This happens when the link appears standing alone in a given line, but line breaks shouldn't be present in rendering at all, unless they were inserted explicitly by double whitespace or `<br/>`.

**Comment by oblitum** [comment]
Actually never mind. This isn't an issue with links in specific at all (but happens more often for me due to them). 

```
❯ cat foo.md
foo
baz
bar

❯ mdcat foo.md
foo
baz
bar
```

`mdcat` is simply rendering newlines, but it shouldn't.

**Comment by lunaryorn** [comment]
That's #4.  Pull requests welcome.  Closing as duplicate.

**Comment by oblitum** [comment]
?

Personally I don't view as a text wrapping problem, I actually don't want text wrapping to any column at all.

**Comment by lunaryorn** [comment]
What _is_ the problem then?  In your second example mdcat appears to render line breaks exactly as they appear in the original markdown file; if neither that nor column wrapping is what you want then what do you want?

In any case my point still stands: Pull requests welcome.  I don't see an obvious bug here myself :shrug:

**Comment by oblitum** [comment]
The bug is to render line breaks exactly as they appear in the original Markdown file. Proper rendering is required to consume them, and it's unrelated text wrapping.

**Comment by lunaryorn** [comment]
I see…

> The bug is to render line breaks exactly as they appear in the original Markdown file.

…so this _is_ issue #4 :slightly_smiling_face: 

> Proper rendering is required to consume them

Says who?  It's not as if there's a law about this…

> it's unrelated to text wrapping

I think it's not.  Someone's got to wrap text if it exceeds the column limit of the terminal window.  If mdcat doesn't the terminal does…

And you want mdcat to discard soft linebreaks, and thus effectively collapse every paragraph into a single long line which makes it a lot more likely that lines exceed the terminal column limit and force the terminal to (poorly) wrap text.

There's nothing for me to fix here.

**Comment by oblitum** [comment]
> …so this _is_ issue #4

No, it's not. Proof: you can have current `mdcat` output wrapped to a given column, and still having no fix on unwanted line breaks, they can still just be there, on top of the additional breaks coming from bare long line wrapping to column.

> Says who? It's not as if there's a law about this…

https://daringfireball.net/projects/markdown/syntax:

> ### Paragraphs and Line Breaks
>
> A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.) Normal paragraphs should not be indented with spaces or tabs.
>
> **The implication of the “one or more consecutive lines of text” rule is that Markdown supports “hard-wrapped” text paragraphs**. This **differs significantly** from most other text-to-HTML formatters (including Movable Type’s “Convert Line Breaks” option) which translate every line break character in a paragraph into a `<br />` tag.
>
> **When you do want to insert a `<br />` break tag using Markdown, you end a line with two or more spaces, then type return**.
>
> Yes, this takes a tad more effort to create a `<br />`, but **a simplistic “every line break is a `<br />`” rule wouldn’t work for Markdown**. Markdown’s email-style blockquoting and multi-paragraph list items work best — and look better — when you format them with hard breaks.

Emphasis mine.

CommonMark follows:

- https://spec.commonmark.org/0.29/#paragraphs
- https://spec.commonmark.org/0.29/#hard-line-break

> I think it's not. Someone's got to wrap text if it exceeds the column limit of the terminal window. If mdcat doesn't the terminal does…
> 
> And you want mdcat to discard soft linebreaks, and thus effectively collapse every paragraph into a single long line which makes it a lot more likely that lines exceed the terminal column limit and force the terminal to (poorly) wrap text.
> 
> I am sorry but there's nothing for me to fix here.

You're mixing two different things and simply denying they're separate. Fixing line breaks doesn't necessarily imply to add text wrapping, it can be done without the latter. Same for the reverse, adding text wrapping doesn't necessarily imply to have those line breaks fixed, it can be done without the latter.

That's why I'm saying it's not the same thing, and I can't be more clearer than this. The Markdown/CommonMark spec is clear, and is what it is. To follow it is what is expected for a software that claims to render it, it's what got to be fixed. It's that, _or at least accept that_, otherwise, You Had One Job™.

I end my case here. I'm using another software.

**Comment by lunaryorn** [comment]
Cool bye
